### PR TITLE
[dev-refactor][HIVE] tc-refactor, scan statistics

### PR DESCRIPTION
**Pull request by Laurawly**
In the progress of refactoring tc, and adding cpu reference code.

### CODE REVIEW

**Comment by neoblizz** [comment]
@Laurawly let me know when the refactor is done and you want me to take a look at it.

**Comment by bkj** [comment]
Does the scan statistics implementation produce the same results as the [Python reference implementation](https://gitlab.hiveprogram.com/ggillary/vertex_nomination_Enron/blob/master/snap_scan_statistics.py)?  I can test if you think it's ready for testing.

File: gunrock/app/ss/ss_enactor.cuh [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+    }
+
+    bool Stop_Condition(int gpu_num = 0)
+    {
+      auto &enactor_slice = this->enactor->enactor_slices[0];
+      auto &enactor_stats = enactor_slice.enactor_stats;
+      auto &data_slice = this->enactor->problem->data_slices[this->gpu_num][0];
+
+      auto &iter = enactor_stats.iteration;
+      if (iter == 1) return true;
```
Comment (neoblizz) [inline gunrock/app/ss/ss_enactor.cuh]: The algorithm only runs for one iteration?

File: gunrock/app/ss/ss_enactor.cuh [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+
+    /**
+      * @brief one run of ss, to be called within GunrockThread
+      * @param thread_data Data for the CPU thread
+      * \return cudaError_t error message(s), if any
+      */
+    cudaError_t Run(ThreadSlice &thread_data)
+    {
+        gunrock::app::Iteration_Loop<
+            // TODO: change to how many {VertexT, ValueT} data need to communicate
```
Comment (neoblizz) [inline gunrock/app/ss/ss_enactor.cuh]: These `TODO` comments can be removed.

File: gunrock/app/tc/tc_test.cuh [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+double CPU_Reference(
+  util::Parameters         &parameters,
+           GraphT          &graph,
+           VertexT         *tc_count)
+{
+    util::CpuTimer cpu_timer;
+    cpu_timer.Start();
+    cpu_timer.Stop();
+    float elapsed = cpu_timer.ElapsedMillis();
+
```
Comment (neoblizz) [inline gunrock/app/tc/tc_test.cuh]: No CPU reference code for TC?

File: gunrock/app/tc/tc_test.cuh [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+typename GraphT::SizeT Validate_Results(
+             util::Parameters &parameters,
+             GraphT           &graph,
+                    VertexT   *h_tc_count,
+                    VertexT   *ref_tc,
+         bool      verbose       = true)
+{
+    typedef typename GraphT::SizeT   SizeT;
+    typedef typename GraphT::CsrT    CsrT;
+
```
Comment (neoblizz) [inline gunrock/app/tc/tc_test.cuh]: Needs either validation of GPU results or CPU reference code to compare against.

**Pull request merged**